cyto_dl/image/transforms/generate_jepa_masks.py

```diff
@@ -0,0 +1,103 @@
+        mask_ratio : float, optional
+            The proportion of the image to be masked
+        """
+        assert mask_ratio < 1, "mask_ratio must be less than 1"
+        assert mask_size * max(block_aspect_ratio) < min(
+            num_patches[-2:]
+        ), "mask_size * max mask aspect ratio must be less than the smallest dimension of num_patches"
+
+        self.mask_size = mask_size
+        self.block_aspect_ratio = block_aspect_ratio
+        self.num_patches = num_patches
+        # convert mask_ratio to number of pixels to be masked
+        self.target_pix = int(mask_ratio * np.prod(num_patches))
+
+        self.mask = np.zeros(num_patches)
+        self.edge_mask = np.ones(num_patches)
+
+        self.spatial_dims = len(num_patches)
+        # create a mask that identified pixels on the edge of the image
+        if self.spatial_dims == 3:
+            self.edge_mask[1:-1, 1:-1, 1:-1] = 0
+        elif self.spatial_dims == 2:
+            self.edge_mask[1:-1, 1:-1] = 0
+        else:
+            raise ValueError("num_patches must be 2 or 3 dimensions")
+
+    def remove_excess_pixels(self, mask):
+        """Remove pixels along the boundary of the mask until the target number of pixels is
+        reached."""
+        bound = find_boundaries(mask, mode="inner")
+        # include image edge as boundary, not just 1:0 transitions
+        edge_mask = np.logical_and(mask, self.edge_mask)
+        bound = np.logical_or(bound, edge_mask)
+        bound_coords = np.argwhere(bound)
+        # find number of pixels to remove from contiguous mask
+        excess = int(mask.sum() - self.target_pix)
+        remove = self.R.choice(range(bound_coords.shape[0]), excess, replace=False)
+        remove_coords = bound_coords[remove]
+        if self.spatial_dims == 3:
+            mask[remove_coords[:, 0], remove_coords[:, 1], remove_coords[:, 2]] = 0
+            mask = rearrange(mask, "z y x -> (z y x)").astype(bool)
+        else:
+            mask[remove_coords[:, 0], remove_coords[:, 1]] = 0
+            mask = rearrange(mask, "y x -> (y x)").astype(bool)
+        return mask
+
+    def __call__(self, img_dict):
+        # generate context (small) and target(large) masks
+        # target: add blocks until target_pix is reached, then randomly remove excess border pixels
+        # context: invert of target mask
+        mask = self.mask.copy()
+        while mask.sum() < self.target_pix:
+            # randomly select block shape
+            aspect_ratio = self.R.uniform(*self.block_aspect_ratio)
+            width = int(self.mask_size * aspect_ratio)
+            height = int(self.mask_size / aspect_ratio)
+            # randomly select block position
+            x = self.R.randint(0, self.num_patches[-1] - width + 1)
+            y = self.R.randint(0, self.num_patches[-2] - height + 1)
+            # add block to mask
+            if self.spatial_dims == 3:
+                mask[:, y : y + height, x : x + width] = 1
```

Comment:
do you not randomly sample dims in Z here?

Reply:
no - I'm basing this off of v-jepa (which has masks that are the same over time) with the idea that 3d images have similar spatial redundancy to natural videos

Reply:
gotcha